xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java

```diff
@@ -399,15 +394,33 @@ private void handleRpcStreamClosed(Status error) {
       rpcRetryTimer = syncContext.schedule(
           new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
 
-      checkArgument(!error.isOk(), "unexpected OK status");
-      String errorMsg = error.getDescription() != null
-          && error.getDescription().equals(CLOSED_BY_SERVER)
-              ? "ADS stream closed with status {0}: {1}. Cause: {2}"
-              : "ADS stream failed with status {0}: {1}. Cause: {2}";
-      logger.log(
-          XdsLogLevel.ERROR, errorMsg, error.getCode(), error.getDescription(), error.getCause());
+      Status newStatus = status;
+      if (responseReceived) {
+        // A closed ADS stream after a successful response is not considered an error. Servers may
+        // close streams for various reasons during normal operation, such as load balancing or
+        // underlying connection hitting its max connection age limit  (see gRFC A9).
+        if (!status.isOk()) {
+          newStatus = Status.OK.withDescription(CLOSED_BY_SERVER);
+          logger.log( XdsLogLevel.INFO, "ADS stream closed with error {0}: {1}. However, a "
+              + "response was received, so this will not be treated as an error. Cause: {2}.",
+              status.getCode(), status.getDescription(), status.getCause());
+        }
+      } else {
+        // If the ADS stream is closed without ever having received a response from the server, then
+        // the XdsClient should consider that a connectivity error (see gRFC A57).
+        if (status.isOk()) {
+          newStatus = Status.UNAVAILABLE.withDescription(
+              "ADS stream failed due to connectivity error");
```

Comment:
It wasn't a connectivity error, it was the server explicitly deciding the close the stream.

Reply:
gRFC says to consider it a connectivity error. Is there any way I can add it in the error as well?

Reply:
"ADS stream closed, per gRFC this is a connectivity error"

Reply:
End user might not be aware of gRFC. How about, "ADS stream failed, because connection was closed before receiving a response"?

Reply:
That sounds good to me.

Reply:
Done